File: src/common/Templates/TemplateHelpers.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections.Generic;

public static class TemplateHelpers
{
    public static string ProcessTemplate(string template, INamedValues values)
    {
        var calculator = new ExpressionCalculator();

        Func<string, string> interpolate = (line) => line.ReplaceValues(values);
        Func<string, bool> evaluateCondition = (condition) =>
        {
            condition = condition.ReplaceValues(values).Trim();
            var evaluated = condition switch
            {
                "" => false,
                "true" => true,
                "false" => false,
                _ => calculator.Evaluate(condition)
            };

            ConsoleHelpers.WriteDebugLine($"Evaluating condition: {condition} => {evaluated}");
            return evaluated;
        };
        Action<string> setVariable = (s) => {
            var result = calculator.Evaluate(s);
            var name = s.Split('=')[0].Trim();
            values.Set(name, result.ToString());
        };
        Func<string, string> include = (nakedAtFile) => AtFileHelpers.ExpandAtFileValue($"@{nakedAtFile}", values);

        return ProcessTemplate(template, evaluateCondition, interpolate, setVariable, include);
    }

    public static string ProcessTemplate(string template, Func<string, bool> evaluateCondition, Func<string, string> interpolate, Action<string> setVariable, Func<string, string> include)
    {
        var safeEvaluateCondition = TryCatchHelpers.NoThrowWrapException(evaluateCondition, false);

        var lines = template.Split('\n').ToList();
        var output = new StringBuilder();

        var inTrueBranchNow = new Stack<bool>();
        inTrueBranchNow.Push(true);

        var skipElseBranches = new Stack<bool>();
        skipElseBranches.Push(true);

        for (var i = 0; i < lines.Count(); i++)
        {
            var line = lines[i];
            var trimmedLine = line.Trim('\n', '\r', ' ', '\t');

            if (trimmedLine.StartsWith("{{if ") && trimmedLine.EndsWith("}}") && !trimmedLine.EndsWith("{{endif}}"))
            {
                var condition = trimmedLine[5..^2].Trim();
                var evaluated = safeEvaluateCondition(condition);
                if (evaluated.ex == null)
                {
                    inTrueBranchNow.Push(evaluated.value);
                    skipElseBranches.Push(evaluated.value);
                    continue;
                }
            }
            else if (trimmedLine.StartsWith("{{else if ") && trimmedLine.EndsWith("}}"))
            {
                if (inTrueBranchNow.Peek())
                {
                    inTrueBranchNow.Pop();
                    inTrueBranchNow.Push(false);
                    // skipElseBranches.Peek() should already be true
                    continue;
                }
                else if (skipElseBranches.Peek())
                {
                    continue;
                }

                var condition = trimmedLine[10..^2].Trim();
                var evaluated = safeEvaluateCondition(condition);
                if (evaluated.ex == null)
                {
                    inTrueBranchNow.Pop();
                    inTrueBranchNow.Push(evaluated.value);
                    skipElseBranches.Pop();
                    skipElseBranches.Push(evaluated.value);
                    continue;
                }
            }
            else if (trimmedLine.StartsWith("{{else}}"))
            {
                if (inTrueBranchNow.Peek())
                {
                    inTrueBranchNow.Pop();
                    inTrueBranchNow.Push(false);
                    // skipElseBranches.Peek() should already be true
                    continue;
                }
                else if (skipElseBranches.Peek())
                {
                    continue;
                }

                inTrueBranchNow.Pop();
                inTrueBranchNow.Push(true);
                skipElseBranches.Pop();
                skipElseBranches.Push(true);
                continue;
            }
            else if (trimmedLine.StartsWith("{{endif}}"))
            {
                inTrueBranchNow.Pop();
                skipElseBranches.Pop();
                continue;
            }
            else if (trimmedLine.StartsWith("{{set ") && trimmedLine.EndsWith("}}"))
            {
                if (inTrueBranchNow.All(b => b))
                {
                    var assignment = trimmedLine[6..^2].Trim();
                    setVariable(assignment);
                }
                continue;
            }
            else if (trimmedLine.StartsWith("{{include ") && trimmedLine.EndsWith("}}"))
            {
                if (inTrueBranchNow.All(b => b))
                {
                    var includeWhat = trimmedLine[10..^2].Trim();
                    var expanded = include(includeWhat)
                        .Split('\n', StringSplitOptions.None)
                        .Select(line => line.TrimEnd('\r').TrimEnd())
                        .ToList();
                    lines.InsertRange(i + 1, expanded);
                }
                continue;
            }
            if (inTrueBranchNow.All(b => b))
            {
                var updated = line.TrimEnd('\n', '\r');

                var firstLine = output.Length == 0;
                if (!firstLine) output.AppendLine();

                var inlineIfPos = updated.IndexOf("{{if ");
                var inlineEndIfPos = updated.IndexOf("{{endif}}");
                var inlineIfEndIf = inlineIfPos >= 0 && inlineEndIfPos >= 0 && inlineIfPos < inlineEndIfPos;
                if (inlineIfEndIf)
                {
                    updated = HandleInlineIf(updated, evaluateCondition, interpolate);
                }

                output.Append(interpolate(updated));
            }
        }

        return output.ToString();
    }

    private static string HandleInlineIf(string line, Func<string, bool> evaluateCondition, Func<string, string> interpolate)
    {
        var safeEvaluateCondition = TryCatchHelpers.NoThrowWrapException(evaluateCondition, false);

        var output = new StringBuilder();

        var inTrueBranchNow = new Stack<bool>();
        inTrueBranchNow.Push(true);

        var skipElseBranches = new Stack<bool>();
        skipElseBranches.Push(true);

        var chars = line.ToCharArray();
        var position = 0;
        while (position < chars.Length)
        {
            var cch = CountCharsToCheck(chars, position);
            if (cch == 1)
            {
                if (inTrueBranchNow.All(b => b))
                {
                    output.Append(chars[position]);
                }
                position++;
                continue;
            }

            var check = new string(chars, position, cch);
            position += cch;

            if (check.StartsWith("{{if ") && check.EndsWith("}}"))
            {
                var condition = check[5..^2].Trim();
                var evaluated = safeEvaluateCondition(condition);
                if (evaluated.ex == null)
                {
                    inTrueBranchNow.Push(evaluated.value);
                    skipElseBranches.Push(evaluated.value);
                    continue;
                }
            }
            else if (check.StartsWith("{{else if ") && check.EndsWith("}}"))
            {
                if (inTrueBranchNow.Peek())
                {
                    inTrueBranchNow.Pop();
                    inTrueBranchNow.Push(false);
                    // skipElseBranches.Peek() should already be true
                    continue;
                }
                else if (skipElseBranches.Peek())
                {
                    continue;
                }

                var condition = check[10..^2].Trim();
                var evaluated = safeEvaluateCondition(condition);
                if (evaluated.ex == null)
                {
                    inTrueBranchNow.Pop();
                    inTrueBranchNow.Push(evaluated.value);
                    skipElseBranches.Pop();
                    skipElseBranches.Push(evaluated.value);
                    continue;
                }
            }
            else if (check.StartsWith("{{else}}") && check.EndsWith("}}"))
            {
                if (inTrueBranchNow.Peek())
                {
                    inTrueBranchNow.Pop();
                    inTrueBranchNow.Push(false);
                    // skipElseBranches.Peek() should already be true
                    continue;
                }
                else if (skipElseBranches.Peek())
                {
                    continue;
                }

                inTrueBranchNow.Pop();
                inTrueBranchNow.Push(true);
                skipElseBranches.Pop();
                skipElseBranches.Push(true);
                continue;
            }
            else if (check.StartsWith("{{endif}}") && check.EndsWith("}}"))
            {
                inTrueBranchNow.Pop();
                skipElseBranches.Pop();
                continue;
            }
            else if (inTrueBranchNow.All(b => b))
            {
                output.Append(interpolate(check));
            }
        }

        return output.ToString();
    }

    private static int CountCharsToCheck(char[] chars, int position)
    {
        if (chars[position] != '{') return 1;

        var cch = 1;
        var braces = 1;
        for (var i = position + 1; i < chars.Length; i++)
        {
            if (chars[i] == '{')
            {
                braces++;
            }
            else if (chars[i] == '}')
            {
                braces--;
                if (braces == 0)
                {
                    cch = i - position + 1;
                    break;
                }
            }
        }

        return cch;
    }
}

File: src/commonjs/Templates/TemplateHelpers.ts
import { INamedValues } from './INamedValues';
import { ExpressionCalculator } from './ExpressionCalculator';
import { TryCatchHelpers } from '../Helpers/TryCatchHelpers';
import { ConsoleHelpers } from '../Helpers/ConsoleHelpers';
import { ValueHelpers } from './ValueHelpers';

// Import the extension methods
import './ValueHelpers';

/**
 * Template processing utilities for conditional logic and variable substitution
 */
export class TemplateHelpers {
    /**
     * Processes a template with the provided values, supporting conditional logic and variable assignment
     * @param template The template string to process
     * @param values The named values for substitution
     * @returns The processed template string
     */
    public static processTemplate(template: string, values: INamedValues): string {
        const calculator = new ExpressionCalculator();

        const interpolate = (line: string): string => line.replaceValues(values);
        const evaluateCondition = (condition: string): boolean => {
            condition = condition.replaceValues(values).trim();
            const evaluated = (() => {
                switch (condition) {
                    case '':
                        return false;
                    case 'true':
                        return true;
                    case 'false':
                        return false;
                    default:
                        return calculator.evaluate(condition) as boolean;
                }
            })();

            ConsoleHelpers.writeDebugLine(`Evaluating condition: ${condition} => ${evaluated}`);
            return evaluated;
        };
        const setVariable = (s: string): void => {
            const result = calculator.evaluate(s);
            const name = s.split('=')[0].trim();
            values.Set(name, result.toString());
        };

        return TemplateHelpers.processTemplateWithCallbacks(template, evaluateCondition, interpolate, setVariable);
    }

    /**
     * Processes a template with custom callback functions
     * @param template The template string to process
     * @param evaluateCondition Function to evaluate boolean conditions
     * @param interpolate Function to interpolate values in strings
     * @param setVariable Function to set variable values
     * @returns The processed template string
     */
    public static processTemplateWithCallbacks(
        template: string,
        evaluateCondition: (condition: string) => boolean,
        interpolate: (line: string) => string,
        setVariable: (assignment: string) => void
    ): string {
        const safeEvaluateCondition = TryCatchHelpers.NoThrowWrapExceptionT(evaluateCondition, false);

        const lines = template.split('\n');
        const output: string[] = [];

        const inTrueBranchNow: boolean[] = [true];
        const skipElseBranches: boolean[] = [true];

        for (const line of lines) {
            const trimmedLine = line.trim();

            if (trimmedLine.startsWith('{{if ') && trimmedLine.endsWith('}}') && !trimmedLine.endsWith('{{endif}}')) {
                const condition = trimmedLine.slice(5, -2).trim();
                const evaluated = safeEvaluateCondition(condition);
                if (evaluated.ex === null) {
                    inTrueBranchNow.push(evaluated.value);
                    skipElseBranches.push(evaluated.value);
                    continue;
                }
            } else if (trimmedLine.startsWith('{{else if ') && trimmedLine.endsWith('}}')) {
                if (inTrueBranchNow[inTrueBranchNow.length - 1]) {
                    inTrueBranchNow.pop();
                    inTrueBranchNow.push(false);
                    // skipElseBranches.peek() should already be true
                    continue;
                } else if (skipElseBranches[skipElseBranches.length - 1]) {
                    continue;
                }

                const condition = trimmedLine.slice(10, -2).trim();
                const evaluated = safeEvaluateCondition(condition);
                if (evaluated.ex === null) {
                    inTrueBranchNow.pop();
                    inTrueBranchNow.push(evaluated.value);
                    skipElseBranches.pop();
                    skipElseBranches.push(evaluated.value);
                    continue;
                }
            } else if (trimmedLine.startsWith('{{else}}')) {
                if (inTrueBranchNow[inTrueBranchNow.length - 1]) {
                    inTrueBranchNow.pop();
                    inTrueBranchNow.push(false);
                    // skipElseBranches.peek() should already be true
                    continue;
                } else if (skipElseBranches[skipElseBranches.length - 1]) {
                    continue;
                }

                inTrueBranchNow.pop();
                inTrueBranchNow.push(true);
                skipElseBranches.pop();
                skipElseBranches.push(true);
                continue;
            } else if (trimmedLine.startsWith('{{endif}}')) {
                inTrueBranchNow.pop();
                skipElseBranches.pop();
                continue;
            } else if (trimmedLine.startsWith('{{set ') && trimmedLine.endsWith('}}')) {
                if (inTrueBranchNow.every(b => b)) {
                    const assignment = trimmedLine.slice(6, -2).trim();
                    setVariable(assignment);
                }
                continue;
            }

            if (inTrueBranchNow.every(b => b)) {
                let updated = line.replace(/\r?\n$/, '');

                const firstLine = output.length === 0;
                if (!firstLine) output.push('');

                const inlineIfPos = updated.indexOf('{{if ');
                const inlineEndIfPos = updated.indexOf('{{endif}}');
                const inlineIfEndIf = inlineIfPos >= 0 && inlineEndIfPos >= 0 && inlineIfPos < inlineEndIfPos;
                if (inlineIfEndIf) {
                    updated = TemplateHelpers.handleInlineIf(updated, evaluateCondition, interpolate);
                }

                output[output.length - 1] = (output[output.length - 1] || '') + interpolate(updated);
            }
        }

        return output.join('\n');
    }

    private static handleInlineIf(
        line: string,
        evaluateCondition: (condition: string) => boolean,
        interpolate: (line: string) => string
    ): string {
        const safeEvaluateCondition = TryCatchHelpers.NoThrowWrapExceptionT(evaluateCondition, false);

        const output: string[] = [];

        const inTrueBranchNow: boolean[] = [true];
        const skipElseBranches: boolean[] = [true];

        const chars = [...line];
        let position = 0;
        while (position < chars.length) {
            const cch = TemplateHelpers.countCharsToCheck(chars, position);
            if (cch === 1) {
                if (inTrueBranchNow.every(b => b)) {
                    output.push(chars[position]);
                }
                position++;
                continue;
            }

            const check = chars.slice(position, position + cch).join('');
            position += cch;

            if (check.startsWith('{{if ') && check.endsWith('}}')) {
                const condition = check.slice(5, -2).trim();
                const evaluated = safeEvaluateCondition(condition);
                if (evaluated.ex === null) {
                    inTrueBranchNow.push(evaluated.value);
                    skipElseBranches.push(evaluated.value);
                    continue;
                }
            } else if (check.startsWith('{{else if ') && check.endsWith('}}')) {
                if (inTrueBranchNow[inTrueBranchNow.length - 1]) {
                    inTrueBranchNow.pop();
                    inTrueBranchNow.push(false);
                    // skipElseBranches.peek() should already be true
                    continue;
                } else if (skipElseBranches[skipElseBranches.length - 1]) {
                    continue;
                }

                const condition = check.slice(10, -2).trim();
                const evaluated = safeEvaluateCondition(condition);
                if (evaluated.ex === null) {
                    inTrueBranchNow.pop();
                    inTrueBranchNow.push(evaluated.value);
                    skipElseBranches.pop();
                    skipElseBranches.push(evaluated.value);
                    continue;
                }
            } else if (check.startsWith('{{else}}') && check.endsWith('}}')) {
                if (inTrueBranchNow[inTrueBranchNow.length - 1]) {
                    inTrueBranchNow.pop();
                    inTrueBranchNow.push(false);
                    // skipElseBranches.peek() should already be true
                    continue;
                } else if (skipElseBranches[skipElseBranches.length - 1]) {
                    continue;
                }

                inTrueBranchNow.pop();
                inTrueBranchNow.push(true);
                skipElseBranches.pop();
                skipElseBranches.push(true);
                continue;
            } else if (check.startsWith('{{endif}}') && check.endsWith('}}')) {
                inTrueBranchNow.pop();
                skipElseBranches.pop();
                continue;
            } else if (inTrueBranchNow.every(b => b)) {
                output.push(interpolate(check));
            }
        }

        return output.join('');
    }

    private static countCharsToCheck(chars: string[], position: number): number {
        if (chars[position] !== '{') return 1;

        let cch = 1;
        let braces = 1;
        for (let i = position + 1; i < chars.length; i++) {
            if (chars[i] === '{') {
                braces++;
            } else if (chars[i] === '}') {
                braces--;
                if (braces === 0) {
                    cch = i - position + 1;
                    break;
                }
            }
        }

        return cch;
    }
}